### ISSUE TITLE
Remove Guard.Assert and throw exceptions manually

### ISSUE THREAD

**Issue by HCanber** [akkadotnet/akka.net#765]
I've looked at how `Guard.Assert` is being used and I don't like it.

Here are a few examples:

``` C#
1. Guard.Assert(!IsEmpty, string.Format("Can't get node for [{0}] from an empty node ring", key));
2. Guard.Assert(TryParseLogMessage(messageStr, out log), "could not parse log message: " + messageStr);
```

Whats wrong? Compare to what they would have looked like without `Guard.Assert`:
``` C#
1. if(!IsEmpty) throw new ArgumentException(string.Format("Can't get node for [{0}] from an empty node ring", key));
2. if(!TryParseLogMessage(messageStr, out log)) throw new ArgumentException("could not parse log message: " + messageStr));
```

In the `Guard.Assert` case a string concatenation and allocation (and in _1._ a `string.Format`) is done __every__ time, compared to only when needed in the if-throw case.

Worse, for _1:_ it shouldn't be an `ArgumentException` at all. It should be `InvalidOperationException`. Not to mention an extra function call is made. In some places of the core framework we don't want that.

I don't want to see this spread.

The scala variant `require` is used for static code analysis tools. That's one reason they're not using if-throw. In scala the expression that's used to create the error message get's compiled into a new class with an `string apply()` function that returns the message, so the expression isn't evaluated until needed.

Do we really need `Guard.Assert`? 
If we keep it where is it ok to use it? 
How do we guarantee that string.Format and string allocations aren't used?

**Comment by rogeralsing** [comment]
I agree, it's better if we can be consistent through the entire code base.
Better to promote correct exception types and consistency over the small benefit of a DbC DSL that is just a few characters shorter than the native boilerplate code.

I did violate this in my old Type.GetType .. != null PR that we reverted for this reason.

**Comment by Aaronontheweb** [comment]
> Do we really need Guard.Assert? 
If we keep it where is it ok to use it? 
How do we guarantee that string.Format and string allocations aren't used? 

`Guard.Assert` is really meant for two things:
1. Future-proofing against standardizing on exception types - i.e. if we want to wrap all of these argument / invalid blah blah exceptions into some type derived from `AkkaException`.
2. Convenience for the developer.

>Better to promote correct exception types and consistency over the small benefit of a DbC DSL that is just a few characters shorter than the native boilerplate code.

I respectfully disagree - teaching every developer what the "correct" exception to throw in each circumstance is non-trivial and neither is showing them how to throw it. It's non-trivial from a code perspective, but actually really substantial from a human knowledge and codebase training perspective. Having everyone default to helper method like `Guard.Assert` actually guarantees greater consistency.

The approach `Guard.Assert` takes is the "right" thing to do when it comes to making life easy for developers, but as you point out it's too simplistic. 

The string concatenations are not a big deal - I just checked all usages `Guard.Assert` in my fork and with the exception of some of the internals for the MultiNodeTestRunner, 100% of `Guard.Assert` calls are made in constructors for objects that aren't called very frequently (`ClusterRouterPool`, `DeadlineFailureDetector`, etc...)

If the issue is that `Guard.Assert` isn't throwing the right type of exception in some cases or isn't doing something that it should, let's fix the implementation.

**Comment by HCanber** [comment]
>1. Future-proofing against standardizing on exception types - i.e. if we want to wrap all of these argument / invalid blah blah exceptions into some type derived from AkkaException.

YAGNI

>2. Convenience for the developer.

Checking for null is easier with Resharper. Three key presses and I have a correctly implemented if-null-throw-ArgumentNullException. So It's not a convenience at every occasion.

>I respectfully disagree - teaching every developer what the "correct" exception to throw in each circumstance is non-trivial and neither is showing them how to throw it

It's the core an inner parts of the Akka framework we're talking about. If devs that contributes don't know how to correctly choose and throw exceptions, which is pretty trivial when put into perspective to what the rest of the code is about , I'm not sure I want those PRs.

>The string concatenations are not a big deal 

Nope, agree. It's easy to fix. The reason I raised this, was that this can become something of an issue. You say that it's hard to teach people to how and what to throw. It's just as hard to teach people how to use Guard.Assert in a way that doesn't affect performance.

>100% of Guard.Assert calls are made in constructors for objects that aren't called very frequently

Right now, yes, but if I understand you correctly you want this everywhere. For example:
``` C#
		protected ImmutableAvlTreeBase(Node node)
		{
			if (node == null) throw new ArgumentNullException("node");
			_node = node;
		}
```
would change to
``` C#
		protected ImmutableAvlTreeBase(Node node)
		{
			Guard.Assert(node != null, "node cannot be null");
			_node = node;
		}
```
which potentially could degrade performance.

**Comment by Aaronontheweb** [comment]
> YAGNI

That's true - this code was ported from Helios originally, before I had a clear idea of what Helios' focus was going to be (this is even before Akka.NET was in the picture) and it was some of the earliest code I wrote for it. At the time I wrote it I wasn't sure if I was going to wrap stuff into `Helios___Exception` or send back a built-in one. 

That's not really an issue in either framework now.

> Nope, agree. It's easy to fix. The reason I raised this, was that this can become something of an issue. You say that it's hard to teach people to how and what to throw. It's just as hard to teach people how to use Guard.Assert in a way that doesn't affect performance.

Ok, I'm sold. Please remove `Guard.Assert` and replace it with manual exception throws.

**Comment by HCanber** [comment]
Changed title from __Uses of Guard.Assert__ to __Remove Guard.Assert and throw exceptions manually__

**Comment by KadekM** [comment]
I made pull request for this here:
https://github.com/akkadotnet/akka.net/pull/785

**Comment by Aaronontheweb** [comment]
Resolved by #785
